flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputChainingOutput.java

```diff
@@ -0,0 +1,138 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.SimpleCounter;
+import org.apache.flink.runtime.metrics.groups.OperatorIOMetricGroup;
+import org.apache.flink.runtime.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.streaming.api.operators.Input;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.metrics.WatermarkGauge;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatusProvider;
+import org.apache.flink.util.OutputTag;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+class MultipleInputChainingOutput<T> implements WatermarkGaugeExposingOutput<StreamRecord<T>> {
+	private static final Logger LOG = LoggerFactory.getLogger(MultipleInputChainingOutput.class);
+
+	protected final Input<T> input;
+	protected final Counter numRecordsIn;
+	protected final WatermarkGauge watermarkGauge = new WatermarkGauge();
+	protected final StreamStatusProvider streamStatusProvider;
+	@Nullable protected final OutputTag<T> outputTag;
+
+	public MultipleInputChainingOutput(
+			Input<T> input,
+			OperatorMetricGroup operatorMetricGroup,
+			StreamStatusProvider streamStatusProvider,
+			@Nullable OutputTag<T> outputTag) {
+		this.input = input;
+
+		{
+			Counter tmpNumRecordsIn;
+			try {
+				OperatorIOMetricGroup ioMetricGroup = operatorMetricGroup.getIOMetricGroup();
+				tmpNumRecordsIn = ioMetricGroup.getNumRecordsInCounter();
+			} catch (Exception e) {
+				LOG.warn("An exception occurred during the metrics setup.", e);
+				tmpNumRecordsIn = new SimpleCounter();
+			}
+			numRecordsIn = tmpNumRecordsIn;
+		}
+
+		this.streamStatusProvider = streamStatusProvider;
+		this.outputTag = outputTag;
+	}
+
+	@Override
+	public void collect(StreamRecord<T> record) {
+		if (this.outputTag != null) {
+			// we are not responsible for emitting to the main output.
+			return;
+		}
+
+		pushToOperator(record);
+	}
+
+	@Override
+	public <X> void collect(OutputTag<X> outputTag, StreamRecord<X> record) {
+		if (this.outputTag == null || !this.outputTag.equals(outputTag)) {
```

Comment:
In that case user should be calling `Output.collect(record)`, as `outputTag` is `@NonNull` (by default).

Reply:
It's unclear to me what is the contract and the intended behavior.
If
> outputTag is @NonNull (by default)

then we probably should check for it.
Currently, we'll just drop the record.

Reply:
If someone would like to pass `outputTag = null`, there is the other `collect(record)` method for that purpose. In Flink we have convention that everything is `@NonNull` unless marked otherwise (or someone forgot about `@Nulalble` annotation). 

Also note that's a pre-existing problem, assuming we can agree that's a problem.

Also, the null check is covered by `org.apache.flink.util.OutputTag#equals` 😸 .

Regardless, I've extracted `OutputTag#isResponsibleFor` method, not sure how worthwhile this was.

Reply:
> In Flink we have convention that everything is @NonNull 

Why there are so many explicit @NonNull and null checks then :) 

> Also, the null check is covered by org.apache.flink.util.OutputTag#equals smile_cat .

It drops the record instead of throwing an exception.

But ok, let's leave it as is.